packages/dds/sequence/src/revertibles.ts

```diff
@@ -0,0 +1,203 @@
+		endSeg,
+		interval.end.getOffset(),
+		ReferenceType.RangeEnd,
+		endSeg.properties,
+	);
+	revertibles.push({
+		event: IntervalEventType.DELETE,
+		interval,
+		start: startRef,
+		end: endRef,
+	});
+
+	return revertibles;
+}
+
+export function appendLocalChangeToRevertibles(
+	string: SharedString,
+	newInterval: SequenceInterval,
+	previousInterval: SequenceInterval,
+	revertibles: IntervalRevertible[],
+) {
+	const startSeg = previousInterval.start.getSegment() as SharedStringSegment;
+	const endSeg = previousInterval.end.getSegment() as SharedStringSegment;
+	const prevStartRef = string.createLocalReferencePosition(
+		startSeg,
+		previousInterval.start.getOffset(),
+		ReferenceType.RangeBegin,
+		startSeg.properties,
+	);
+	const prevEndRef = string.createLocalReferencePosition(
+		endSeg,
+		previousInterval.end.getOffset(),
+		ReferenceType.RangeEnd,
+		endSeg.properties,
+	);
+	revertibles.push({
+		event: IntervalEventType.CHANGE,
+		interval: newInterval,
+		start: prevStartRef,
+		end: prevEndRef,
+	});
+
+	return revertibles;
+}
+
+export function appendLocalPropertyChangedToRevertibles(
+	interval: SequenceInterval,
+	deltas: PropertySet,
+	revertibles: IntervalRevertible[],
+) {
+	revertibles.push({
+		event: IntervalEventType.PROPERTYCHANGED,
+		interval,
+		propertyDeltas: deltas,
+	});
+
+	return revertibles;
+}
+// Uses of referenceRangeLabels will be removed once AB#4081 is completed.
+function revertLocalAdd(
+	string: SharedString,
+	revertible: TypedRevertible<typeof IntervalEventType.ADD>,
+) {
+	// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+	const id = revertible.interval.getIntervalId()!;
+	const label = revertible.interval.properties.referenceRangeLabels[0];
+	string.getIntervalCollection(label).removeIntervalById(id);
+}
+
+function revertLocalDelete(
+	string: SharedString,
+	revertible: TypedRevertible<typeof IntervalEventType.DELETE>,
+) {
+	const label = revertible.interval.properties.referenceRangeLabels[0];
+	const start = string.localReferencePositionToPosition(revertible.start);
+	const end = string.localReferencePositionToPosition(revertible.end);
+	const type = revertible.interval.intervalType;
+	const props = revertible.interval.properties;
+	string.getIntervalCollection(label).add(start, end, type, props);
+
+	string.removeLocalReferencePosition(revertible.start);
+	string.removeLocalReferencePosition(revertible.end);
+}
+
+function revertLocalChange(
+	string: SharedString,
+	revertible: TypedRevertible<typeof IntervalEventType.CHANGE>,
+) {
+	const label = revertible.interval.properties.referenceRangeLabels[0];
+	// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+	const id = revertible.interval.getIntervalId()!;
+	const start = string.localReferencePositionToPosition(revertible.start);
+	const end = string.localReferencePositionToPosition(revertible.end);
+	string.getIntervalCollection(label).change(id, start, end);
```

Comment:
It's arguable what should happen if a remote move/change happened in the meantime. E.g. The interval is at the beginning of the string, I moved the interval to the middle of the string, a remote client moves it to the end, I revert my edit. Similarly for property changes. We could handle this, but I'm inclined to keep the code simple the way it is here. What do you think?

Reply:
overwriting the remote edit is consistent with the analogous conflicts you mentioned for property changes on MergeTree segments (though by "property changes" I guess you could mean on intervals--either way similar case). Keeping it simple is fine imo.